lib/icinga/dependency-group.cpp

```diff
@@ -0,0 +1,362 @@
+ * For explicitly configured redundancy groups, the name of the dependency group is the same as the one located
+ * in the configuration files. For non-redundant dependencies, on the other hand, the name is always empty.
+ *
+ * @return - Returns the name of the current dependency group.
+ */
+const String& DependencyGroup::GetName() const
+{
+	// We don't need to lock the mutex here, as the name is set once during
+	// the object construction and never changed afterwards.
+	return m_Name;
+}
+
+/**
+ * Retrieve the unique composite key of the current dependency group.
+ *
+ * The composite key consists of some unique data of the group members, and should be used to generate
+ * a unique deterministic hash for the dependency group. Additionally, for explicitly configured redundancy
+ * groups, the non-unique dependency group name is also included on top of the composite keys.
+ *
+ * @return - Returns the composite key of the current dependency group.
+ */
+String DependencyGroup::GetCompositeKey()
+{
+	std::lock_guard lock(m_Mutex);
+	if (m_CompositeKey.IsEmpty()) {
+		Array::Ptr data(new Array{GetName()});
+		for (auto& [compositeKey, _] : m_Members) {
+			auto [parent, tp, stateFilter, ignoreSoftStates] = compositeKey;
+			data->Add(std::move(parent));
+			data->Add(std::move(tp));
+			data->Add(stateFilter);
+			data->Add(ignoreSoftStates);
+		}
+
+		m_CompositeKey = PackObject(data);
+	}
+
+	return m_CompositeKey;
+}
+
+/**
+ * Retrieve the state of the current dependency group.
+ *
+ * The state of the dependency group is determined based on the state of the members of the group.
+ * A dependency group is considered unreachable when none of the members is reachable. However, a
+ * dependency group may still be marked a failed even when it has reachable members, but an unreachable
+ * group has always a failed state.
+ *
+ * @return - Returns the state of the current dependency group.
+ */
+DependencyGroup::State DependencyGroup::GetState(DependencyType dt, int rstack) const
+{
+	MembersMap members;
+	{
+		// We don't want to hold the mutex lock for the entire evaluation, thus we just need to operate on a copy.
+		std::lock_guard lock(m_Mutex);
+		members = m_Members;
+	}
+
+	State state{false /* Reachable */, false /* OK */};
+	for (auto& [_, dependencies] : members) {
+		for (auto& [checkable, dependency] : dependencies) {
+			state.Reachable = dependency->GetParent()->IsReachable(dt, rstack);
+			if (!state.Reachable && !IsRedundancyGroup()) {
+				return state;
+			}
+
+			if (state.Reachable) {
+				state.OK = dependency->IsAvailable(dt);
+				// If this is a redundancy group, and we have found one functional path, that's enough and we can return.
+				// Likewise, if this is a non-redundant dependency group, and we have found one non-functional path,
+				// we have to mark the group as failed and return.
+				if (state.OK == IsRedundancyGroup()) { // OK && IsRedundancyGroup() || !OK && !IsRedundancyGroup()
+					return state;
+				}
```

Comment:
That seems to always consider the dependency. Shouldn't there also be a case that can skip a specific dependency when it's outside of its configures time period?

Reply:
The previous implementation had no such special handling, so I won't do that here as well, but if in doubt, will fix it in a separate PR as discussed.

Reply:
I've created #10377 for that, which also includes additional information.